newton/tests/test_viewer_world_offsets.py

```diff
@@ -0,0 +1,268 @@
+        self.assertIsNotNone(world_instance, "World instance not found")
+
+        # Update transforms
+        viewer.begin_frame(0.0)
+        ground_instance.update(state, world_offsets=viewer.world_offsets)
+        world_instance.update(state, world_offsets=viewer.world_offsets)
+
+        # Check ground plane is at origin (unaffected by offsets)
+        ground_xform = ground_instance.world_xforms.numpy()[0]
+        assert_np_equal(ground_xform[:3], np.array([0.0, 0.0, 0.0]), tol=1e-5)
+
+        # Check world shapes are offset
+        world_xforms = world_instance.world_xforms.numpy()
+        expected_offsets = np.array([[-5.0, 0.0, 1.0], [5.0, 0.0, 1.0]])
+
+        for i in range(num_worlds):
+            assert_np_equal(world_xforms[i][:3], expected_offsets[i], tol=1e-5)
+
+
+def test_visual_separation(test: TestViewerWorldOffsets, device):
+    """Test that viewer offsets provide visual separation without affecting physics."""
+    num_worlds = 4
+    spacing = (5.0, 5.0, 0.0)
+
+    # Create model
+    builder = newton.ModelBuilder()
+    world = newton.ModelBuilder()
+    world.add_body(
+        xform=wp.transform(wp.vec3(0.0, 0.0, 1.0), wp.quat_identity()),
+        mass=1.0,
+        I_m=wp.mat33(np.eye(3)),
+        key="test_body",
+    )
+    cfg = newton.ModelBuilder.ShapeConfig(density=1000.0)
+    world.add_shape(
+        body=0,  # Attach to the first (and only) body in world
+        type=newton.GeoType.BOX,
+        scale=wp.vec3(0.5, 0.5, 0.5),
+        cfg=cfg,
+    )
+
+    builder.replicate(world, num_worlds)
+    model = builder.finalize(device=device)
+    state = model.state()
+
+    # Create viewer and set offsets
+    viewer = ViewerNull(num_frames=1)
+    viewer.set_model(model)
+    viewer.set_world_offsets(num_worlds, spacing)
+
+    # Get shape instances from viewer
+    shape_instances = next(iter(viewer._shape_instances.values()))
+
+    # Update transforms
+    viewer.begin_frame(0.0)
+    shape_instances.update(state, world_offsets=viewer.world_offsets)
+
+    # Check that world transforms have been offset
+    world_xforms = shape_instances.world_xforms.numpy()
+
+    # Expected offsets based on 2x2 grid with spacing (5, 5, 0)
+    expected_offsets = np.array(
+        [
+            [-2.5, -2.5, 0.0],  # env 0
+            [-2.5, 2.5, 0.0],  # env 1
+            [2.5, -2.5, 0.0],  # env 2
+            [2.5, 2.5, 0.0],  # env 3
+        ]
+    )
+
+    for i in range(num_worlds):
+        actual_pos = world_xforms[i][:3]
+        expected_pos = expected_offsets[i] + np.array([0.0, 0.0, 1.0])  # body is at (0,0,1)
+        assert_np_equal(actual_pos, expected_pos, tol=1e-4)
+
+
+# Add device-specific tests
+devices = get_test_devices()
+for device in devices:
+    add_function_test(
+        TestViewerWorldOffsets,
+        f"test_visual_separation_{device.alias}",
+        test_visual_separation,
+        devices=[device],
+    )
+
+
+if __name__ == "__main__":
+    wp.clear_kernel_cache()
```

Comment:
I think it's better to remove this clear cache function here, we shouldn't need it.